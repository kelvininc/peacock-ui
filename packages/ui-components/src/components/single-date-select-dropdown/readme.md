# kv-single-date-select-dropdown

<!-- Auto Generated Below -->


## Usage

### Angular

```html
<!-- Default -->
<kv-single-date-select-dropdown></kv-single-date-select-dropdown>

<!-- With input config -->
<kv-single-date-select-dropdown [dropdownConfig]="{ label: 'Birthday', placeholder: 'Select a date' }"></kv-single-date-select-dropdown>
```


### React

```tsx
import React from 'react';

import { KvSingleDateSelectDropdown } from '@kelvininc/react-ui-components';

export const KvSingleDateSelectDropdownExample: React.FC = () => (
	<>
		{/*-- Default --*/}
		<KvSingleDateSelectDropdown />

		{/*-- With input config --*/}
		<KvSingleDateSelectDropdown dropdownConfig={{ label: 'Birthday', placeholder: 'Select a date' }} />
	</>
);
```



## Properties

| Property        | Attribute       | Description                                                   | Type                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | Default     |
| --------------- | --------------- | ------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| `dateMask`      | `date-mask`     | (optional) The text field date format mask in ISO 8601 format | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | `undefined` |
| `disabledDates` | --              | (options) The disabled dates array                            | `string[]`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | `[]`        |
| `initialDate`   | `initial-date`  | (optional) Initial date                                       | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | `undefined` |
| `inputConfig`   | --              | (optional) The text field dropdown custom configurations      | `{ type?: EInputFieldType; label?: string; icon?: EIconName \| EOtherIconName; inputName?: string; examples?: string[]; placeholder?: string; maxLength?: number; minLength?: number; max?: string \| number; min?: string \| number; step?: string \| number; size?: EComponentSize; disabled?: boolean; required?: boolean; loading?: boolean; state?: EValidationState; helpText?: string \| string[]; value?: string \| number; readonly?: boolean; forcedFocus?: boolean; tooltipConfig?: Partial<ITooltip>; }` | `{}`        |
| `maxDate`       | `max-date`      | (optional) Maximum accepted date                              | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | `undefined` |
| `minDate`       | `min-date`      | (optional) Minimum accepted date                              | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | `undefined` |
| `selectedDate`  | `selected-date` | (optional) Currently selected date                            | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | `undefined` |


## Events

| Event             | Description                                            | Type                       |
| ----------------- | ------------------------------------------------------ | -------------------------- |
| `openStateChange` | Emitted when the calendar selector opens state changes | `CustomEvent<boolean>`     |
| `selectDate`      | Emitted when day is selected                           | `CustomEvent<ISelectDate>` |


## Dependencies

### Depends on

- [kv-dropdown](../dropdown)
- [kv-calendar-single-date-selector](../calendar-single-date-selector)

### Graph
```mermaid
graph TD;
  kv-single-date-select-dropdown --> kv-dropdown
  kv-single-date-select-dropdown --> kv-calendar-single-date-selector
  kv-dropdown --> kv-dropdown-base
  kv-dropdown --> kv-text-field
  kv-dropdown --> kv-icon
  kv-text-field --> kv-tooltip
  kv-text-field --> kv-form-label
  kv-text-field --> kv-icon
  kv-text-field --> kv-form-help-text
  kv-form-help-text --> kv-icon
  kv-calendar-single-date-selector --> kv-calendar
  kv-calendar --> kv-action-button-icon
  kv-calendar --> kv-calendar-day
  kv-action-button-icon --> kv-action-button
  kv-action-button-icon --> kv-icon
  kv-action-button-icon --> kv-badge
  style kv-single-date-select-dropdown fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


