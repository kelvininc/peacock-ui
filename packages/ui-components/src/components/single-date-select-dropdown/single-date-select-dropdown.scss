@import '../../assets/styles/globals';

:host {
	.single-date-select-dropdown {
		position: relative;
		user-select: none;
	}

	.calendar-container {
		position: absolute;
		z-index: 2;
	}
}
