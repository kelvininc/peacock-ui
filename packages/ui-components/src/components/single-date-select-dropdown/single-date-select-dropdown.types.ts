import { EventEmitter } from '@stencil/core';
import { ITextField } from '../../types';
import { ICalendarSingleDateSelector, ICalendarSingleDateSelectorEvents } from '../calendar-single-date-selector/calendar-single-date-selector.types';

export interface ISingleDateSelectDropdown extends ICalendarSingleDateSelector {
	/** (optional) The text field dropdown custom configurations */
	inputConfig?: Partial<ITextField>;
	/** (optional) The text field date format mask in ISO 8601 format */
	dateMask?: string;
}

export interface ISingleDateSelectDropdownEvents extends ICalendarSingleDateSelectorEvents {
	/** Emitted when the calendar selector opens state changes */
	openStateChange: EventEmitter<boolean>;
}
