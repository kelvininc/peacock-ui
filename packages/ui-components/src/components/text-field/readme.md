# _<kv-text-field>_

<!-- Auto Generated Below -->


## Usage

### Angular / javascript

```html
<!-- Default -->
<kv-text-field></kv-text-field>

<!-- Labeled -->
<kv-text-field label="Text Field"></kv-text-field>

<!-- Disabled -->
<kv-text-field disabled></kv-text-field>
```


### React

```tsx
import React from 'react';

import { KvTextField } from '@kelvininc/react-ui-components';

export const TextFieldExample: React.FC = () => (
	<>
		{/*-- Default --*/}
		<KvTextField />

		{/*-- Labeled --*/}
		<KvTextField label="Text Field" />

		{/*-- Disabled --*/}
		<KvTextField disabled />
	</>
);

```


### Stencil

```tsx
import { Component, h } from '@stencil/core';

@Component({
	tag: 'text-field-example',
	styleUrl: 'text-field-example.css',
	shadow: true,
})
export class TextFieldExample {
	render() {
		return [
			// Default
			<kv-text-field></kv-text-field>

			// Labeled
			<kv-text-field label="Text Field"></kv-text-field>

			// Disabled
			<kv-text-field disabled></kv-text-field>
		];
	}
}

```



## Properties

| Property         | Attribute          | Description                                                                        | Type                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | Default                 |
| ---------------- | ------------------ | ---------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------- |
| `actionIcon`     | `action-icon`      | (optional) Icon that is added on the right of the input. Its clickable.            | `EIconName \| EOtherIconName`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | `undefined`             |
| `badge`          | `badge`            | (optional) Text to display inside a badge on the right side of the displayed value | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | `undefined`             |
| `disabled`       | `disabled`         | (optional) Text field disabled                                                     | `boolean`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `false`                 |
| `examples`       | --                 | (optional) Text field example values                                               | `string[]`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | `undefined`             |
| `forcedFocus`    | `forced-focus`     | (optional) Text field focus state                                                  | `boolean`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `false`                 |
| `helpText`       | `help-text`        | (optional) Text field help text                                                    | `string \| string[]`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | `[]`                    |
| `icon`           | `icon`             | (optional) Text field's icon symbol name                                           | `EIconName \| EOtherIconName`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | `undefined`             |
| `inputMaskRegex` | `input-mask-regex` | (optional) Input mask regex                                                        | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | `''`                    |
| `inputName`      | `input-name`       | (optional) Text field input name                                                   | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | `undefined`             |
| `label`          | `label`            | (optional) Text field label                                                        | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | `undefined`             |
| `loading`        | `loading`          | (optional) Text field loading state                                                | `boolean`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `false`                 |
| `max`            | `max`              | (optional) Text field maximum value                                                | `number \| string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | `undefined`             |
| `maxLength`      | `max-length`       | (optional) Text field maximum number of characters required                        | `number`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | `undefined`             |
| `min`            | `min`              | (optional) Text field minimum value                                                | `number \| string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | `undefined`             |
| `minLength`      | `min-length`       | (optional) Text field minimum number of characters required                        | `number`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | `undefined`             |
| `placeholder`    | `placeholder`      | (optional) Text field place holder                                                 | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | `''`                    |
| `readonly`       | `readonly`         | (optional) Text field is readonly                                                  | `boolean`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `false`                 |
| `required`       | `required`         | (optional) Text field required                                                     | `boolean`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `false`                 |
| `size`           | `size`             | (optional) Sets this tab item to a different styling configuration                 | `EComponentSize.Large \| EComponentSize.Small`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     | `EComponentSize.Large`  |
| `state`          | `state`            | (optional) Text field state                                                        | `EValidationState.Invalid \| EValidationState.None \| EValidationState.Valid`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | `EValidationState.None` |
| `step`           | `step`             | (optional) Text field interval between legal numbers                               | `number \| string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | `undefined`             |
| `tooltipConfig`  | --                 | (optional) Text field tooltip configuration                                        | `{ text?: string; position?: ETooltipPosition; allowedPositions?: ETooltipPosition[]; options?: Partial<{ placement?: Placement; strategy?: Strategy; middleware?: (false \| { name: string; options?: any; fn: (state: { x: number; y: number; initialPlacement: Placement; platform: Platform; placement: Placement; strategy: Strategy; middlewareData: MiddlewareData; rects: ElementRects; elements: Elements; }) => Promisable<MiddlewareReturn>; })[]; platform?: Platform; }>; disabled?: boolean; contentElement?: HTMLElement; truncate?: boolean; delay?: number; withArrow?: boolean; customClass?: CustomCssClass; }` | `undefined`             |
| `type`           | `type`             | (optional) Text field type                                                         | `EInputFieldType.Date \| EInputFieldType.DateTime \| EInputFieldType.Email \| EInputFieldType.Number \| EInputFieldType.Password \| EInputFieldType.Radio \| EInputFieldType.Text`                                                                                                                                                                                                                                                                                                                                                                                                                                                 | `EInputFieldType.Text`  |
| `useInputMask`   | `use-input-mask`   | (optional) Use a input mask when the text field type is number (default true)      | `boolean`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | `false`                 |
| `value`          | `value`            | (optional) Text field value                                                        | `number \| string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | `''`                    |
| `valuePrefix`    | `value-prefix`     | (optional) Defines the prefix that adds context to displayed values                | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           | `undefined`             |


## Events

| Event              | Description                                  | Type                      |
| ------------------ | -------------------------------------------- | ------------------------- |
| `fieldClick`       | Emmited when there's a click on this element | `CustomEvent<MouseEvent>` |
| `rightActionClick` | Emitted when the right icon is clicked       | `CustomEvent<MouseEvent>` |
| `textChange`       | Emitted when a keyboard input occurred       | `CustomEvent<string>`     |
| `textFieldBlur`    | Emitted when text field lost focus           | `CustomEvent<string>`     |


## CSS Custom Properties

| Name                                | Description                                    |
| ----------------------------------- | ---------------------------------------------- |
| `--background-color-default`        | Background color when state is default.        |
| `--background-color-disabled`       | Background color when state is disabled.       |
| `--border-color-default`            | Border color when state is default.            |
| `--border-color-error`              | Border color when state is invalid.            |
| `--border-color-focused`            | Border color when state is focused.            |
| `--input-height-large`              | Text Field's large height.                     |
| `--input-height-small`              | Text Field's small height.                     |
| `--input-max-width`                 | Text Field's max width.                        |
| `--input-min-width`                 | Text Field's min width.                        |
| `--input-width`                     | Text Field's width.                            |
| `--text-color-help-text-default`    | Help Text color when state is default.         |
| `--text-color-help-text-error`      | Help Text color when state is invalid.         |
| `--text-color-icon-default`         | Icon color when state is default.              |
| `--text-color-icon-disabled`        | Icon color when state is disabled.             |
| `--text-color-input-default`        | Input text color when state is default.        |
| `--text-color-input-disabled`       | Input text color when state is disabled.       |
| `--text-color-input-focused`        | Input Text color when state is focused.        |
| `--text-color-label`                | Label Text color.                              |
| `--text-color-placeholder-default`  | Placeholder text color when state is default.  |
| `--text-color-placeholder-disabled` | Placeholder text color when state is disabled. |
| `--text-color-placeholder-focused`  | Placeholder text color when state is focused.  |


## Dependencies

### Used by

 - [kv-dropdown](../dropdown)
 - [kv-range-dates-select-dropdown](../range-dates-select-dropdown)
 - [kv-search](../search)

### Depends on

- [kv-tooltip](../tooltip)
- [kv-form-label](../form-label)
- [kv-icon](../icon)
- [kv-badge](../badge)
- [kv-form-help-text](../form-help-text)

### Graph
```mermaid
graph TD;
  kv-text-field --> kv-tooltip
  kv-text-field --> kv-form-label
  kv-text-field --> kv-icon
  kv-text-field --> kv-badge
  kv-text-field --> kv-form-help-text
  kv-tooltip --> kv-portal
  kv-tooltip --> kv-tooltip-text
  kv-form-help-text --> kv-icon
  kv-dropdown --> kv-text-field
  kv-range-dates-select-dropdown --> kv-text-field
  kv-search --> kv-text-field
  style kv-text-field fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


