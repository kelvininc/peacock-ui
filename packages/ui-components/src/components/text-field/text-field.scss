@import '../../assets/styles/globals';

$default-input-height: var(--input-height, $spacing-9x);
$slim-input-height: var(--input-height, $spacing-7x);

.label,
.error {
	@include kv-font-label-small-uppercase-semibold; }

.label {
	color: kv-color('text');
}

.error {
	color: kv-color('error');
	margin-right: $spacing;
}

.flex-align-center {
	display: flex;
	align-content: center;
}

.text-container {
	margin-left: $spacing-2x;

	.text-elem {
		@include kv-font-label-small-regular;
		width: 100%;
		height: 16px;
		margin-top: $spacing;

		color: kv-color('neutral-5');

		.flex-align-center {
			display: flex;
			align-content: center;
		}

		svg-icon {
			margin-right: $spacing;
		}
	}

	.has-error {
		color: kv-color('error');
	}
}

input {
	@include kv-font-span-regular;

	color: kv-color('primary');
	width: calc(100% - $spacing-4x);
	padding-left: $spacing-2x;
	padding-right: $spacing-2x;
	padding-top: 0;
	padding-bottom: 0;
	outline: 0;
	border-radius: 4px;
	border: solid 1px kv-color('neutral-2');
	height: $default-input-height;

	.has-error {
		border-color: kv-color('error');
	}

	&:disabled {
		background-color: kv-color('neutral-3');
		color: kv-color('text');
	}

	&::placeholder {
		color: kv-color('neutral-2');
	}

	&:disabled::placeholder {
		color: kv-color('neutral-5');
	}

	&:focus {
		border-color: kv-color('primary');
		color: kv-color('primary');
	}
}

.input-container {
	position: relative;
	margin-top: $spacing;

	.svg-icon {
		position: absolute;
		top: 0;
		right: $spacing-2x;
		height: 100%;
		cursor: pointer;
		user-select: none;
	}
}

.input-container-loading {

	@include ghost-animation;
	display: flex;
	height: $default-input-height;
	margin-top: $spacing;
}

.kv-slim-text-field {
	height: $slim-input-height;
}
