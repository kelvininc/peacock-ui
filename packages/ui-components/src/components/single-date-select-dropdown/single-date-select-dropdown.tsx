import { Component, Event, EventEmitter, Host, Prop, State, h } from '@stencil/core';
import { ISelectDate, ITextField } from '../../types';
import { ISingleDateSelectDropdown, ISingleDateSelectDropdownEvents } from './single-date-select-dropdown.types';
import { isEmpty, merge } from 'lodash-es';

import { DEFAULT_DATE_INPUT_CONFIG } from './single-date-select-dropdown.config';
import { formatDateTime } from '../../utils/date.helper';

@Component({
	tag: 'kv-single-date-select-dropdown',
	styleUrl: 'single-date-select-dropdown.scss',
	shadow: true
})
export class KvSingleDateSelectDropdown implements ISingleDateSelectDropdown, ISingleDateSelectDropdownEvents {
	/** @inheritdoc */
	@Prop({ reflect: false }) inputConfig?: Partial<ITextField> = {};
	/** @inheritdoc */
	@Prop({ reflect: true }) selectedDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) initialDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) disabledDates?: string[] = [];
	/** @inheritdoc */
	@Prop({ reflect: false }) minDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) maxDate?: string;
	/** @inheritdoc */
	@Prop({ reflect: false }) dateMask?: string;

	/** @inheritdoc */
	@Event({ bubbles: false }) openStateChange: EventEmitter<boolean>;
	/** @inheritdoc */
	@Event() selectDate: EventEmitter<ISelectDate>;

	@State() isDropdownOpen: boolean = false;

	private onDropdownChange = ({ detail: openState }: CustomEvent<boolean>) => {
		this.isDropdownOpen = openState;
	};

	private getInputConfig = (): Partial<ITextField> => {
		return merge({}, DEFAULT_DATE_INPUT_CONFIG, this.inputConfig, { value: this.getFormattedSelectedDate() });
	};

	public getFormattedSelectedDate = (): string | undefined => {
		if (!isEmpty(this.selectedDate)) {
			return formatDateTime(this.selectedDate, this.dateMask);
		}
	};

	render() {
		return (
			<Host>
				<div class="single-date-select-dropdown">
					<kv-dropdown isOpen={this.isDropdownOpen} onOpenStateChange={this.onDropdownChange} inputConfig={this.getInputConfig()}>
						<div class="calendar-container">
							<kv-calendar-single-date-selector
								selectedDate={this.selectedDate}
								initialDate={this.initialDate}
								disabledDates={this.disabledDates}
								minDate={this.minDate}
								maxDate={this.maxDate}
							/>
						</div>
					</kv-dropdown>
				</div>
			</Host>
		);
	}
}
