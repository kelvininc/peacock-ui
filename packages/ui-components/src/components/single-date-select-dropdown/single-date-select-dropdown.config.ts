import { EIconName } from '../icon/icon.types';
import { EInputFieldType, EValidationState, ITextField } from '../text-field/text-field.types';

export const DEFAULT_DATE_INPUT_CONFIG: Partial<ITextField> = {
	placeholder: 'Select a date',
	type: EInputFieldType.Text,
	icon: EIconName.Calendar,
	state: EValidationState.None
};
