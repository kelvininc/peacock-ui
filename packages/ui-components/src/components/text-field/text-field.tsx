import { Component, Element, Event, EventEmitter, Fragment, h, Host, Prop, State, Watch } from '@stencil/core';
import { isEmpty, isNil } from 'lodash-es';
import Inputmask from 'inputmask';
import { EComponentSize, ETooltipPosition } from '../../utils/types';
import { EInputFieldType, EValidationState, ITextFieldEvents, ITextField } from './text-field.types';
import { EIconName, EOtherIconName } from '../icon/icon.types';
import { NUMERIC_TEXT_INPUT_MASK_CONFIG } from './text-field.config';

@Component({
	tag: 'kv-text-field',
	styleUrls: {
		night: 'text-field.night.scss',
		light: 'text-field.light.scss'
	},
	shadow: true
})
export class KvTextField implements ITextField, ITextFieldEvents {
	private nativeInput?: HTMLInputElement;

	@Element() el!: HTMLKvTextFieldElement;
	/** @inheritdoc */
	@Prop({ reflect: true }) type: EInputFieldType = EInputFieldType.Text;
	/** @inheritdoc */
	@Prop({ reflect: true }) label?: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) examples?: string[];
	/** @inheritdoc */
	@Prop({ reflect: true }) icon?: EIconName | EOtherIconName;
	/** @inheritdoc */
	@Prop({ reflect: true }) inputName?: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) placeholder?: string;
	/** @inheritdoc */
	@Prop({ reflect: true }) maxLength?: number;
	/** @inheritdoc */
	@Prop({ reflect: true }) minLength?: number;
	/** @inheritdoc */
	@Prop({ reflect: true }) max?: string | number;
	/** @inheritdoc */
	@Prop({ reflect: true }) min?: string | number;
	/** @inheritdoc */
	@Prop({ reflect: true }) step?: string | number;
	/** @inheritdoc */
	@Prop() size: EComponentSize = EComponentSize.Large;
	/** @inheritdoc */
	@Prop({ reflect: true }) disabled: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) required: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) loading: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) state: EValidationState = EValidationState.None;
	/** @inheritdoc */
	@Prop({ reflect: true }) uneditable: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) helpText: string | string[] = [];
	/** @inheritdoc */
	@Prop({ reflect: true }) forcedFocus: boolean = false;
	/** @inheritdoc */
	@Prop({ reflect: true }) tooltip?: string;
	/** @inheritdoc */
	@Prop({ reflect: true, mutable: true }) value?: string | number | null = '';
	/** Watch `value` property for changes and update native input element accordingly */
	@Watch('value')
	valueChangeHandler(newValue: string | number | null) {
		this.value = newValue;

		const nativeInput = this.nativeInput;
		const value = this.getValue();

		if (nativeInput && nativeInput.value !== value) {
			/**
			 * Assigning the native input's value on attribute
			 * value change, allows `textChange` implementations
			 * to override the control's value.
			 *
			 * Used for patterns such as input trimming (removing whitespace),
			 * or input masking.
			 */
			nativeInput.value = value;
		}
	}

	/** Internal help texts state */
	@State() _helpTexts: string[];
	/** Watch the `helpText` property and update internal state accordingly */
	@Watch('helpText')
	helpTextChangeHandler(newValue: string | string[]) {
		this._helpTexts = this.buildHelpTextMessages(newValue);
	}
	/** Watch the `forcedFocus` property and update internal state accordingly */
	@Watch('forcedFocus')
	forcedFocusChangeHandler(newValue: boolean) {
		this.focused = newValue;

		if (!this.focused) {
			this.el.blur();
		}
	}

	componentWillLoad() {
		// Init the states because Watches run only on component updates
		this._helpTexts = this.buildHelpTextMessages(this.helpText);
		this.focused = this.forcedFocus;
	}

	componentDidLoad() {
		if (this.type === EInputFieldType.Number) {
			return Inputmask({
				...NUMERIC_TEXT_INPUT_MASK_CONFIG,
				min: this.min,
				max: this.max
			}).mask(this.nativeInput);
		}
	}

	/** Text field focus state */
	@State() focused = false;

	/** @inheritdoc */
	@Event() textChange: EventEmitter<string>;
	/** @inheritdoc */
	@Event() textFieldBlur: EventEmitter<string>;

	private onInputHandler = ({ target }: InputEvent) => {
		const input = target as HTMLInputElement | null;
		if (!isNil(input)) {
			this.value = input.value || '';
		}
		this.textChange.emit(this.getValue());
	};

	private onBlurHandler = ({ target }: FocusEvent) => {
		if (this.forcedFocus) {
			return;
		}

		this.textFieldBlur.emit((target as HTMLInputElement).value);
		this.focused = false;
	};

	private onFocusHandler = () => {
		this.focused = true;
	};

	private buildHelpTextMessages(value: string | string[]) {
		value = value || [];
		return Array.isArray(value) ? value : [value];
	}

	private get hasRightSlot() {
		return !isNil(this.el.querySelector('[slot="right-slot"]'));
	}

	private getValue(): string {
		return typeof this.value === 'number' ? this.value.toString() : (this.value || '').toString();
	}

	private getType(): string {
		return this.type === EInputFieldType.Number ? EInputFieldType.Text : this.type;
	}

	render() {
		const id = this.el.getAttribute('id');
		const value = this.getValue();
		const type = this.getType();

		return (
			<Host>
				<kv-tooltip text={this.tooltip} position={ETooltipPosition.TopStart}>
					<div class="text-field-container">
						<kv-form-label label={this.label} required={this.required}></kv-form-label>
						<div
							class={{
								'input-container': true,
								[`input-container--size-${this.size}`]: true
							}}
						>
							{!this.loading && (
								<Fragment>
									<input
										id={id}
										ref={input => (this.nativeInput = input)}
										type={type}
										list={!isNil(this.examples) ? `examples_${id}` : undefined}
										name={this.inputName}
										placeholder={this.placeholder}
										disabled={this.disabled}
										max={this.max}
										min={this.min}
										maxLength={this.maxLength}
										minLength={this.minLength}
										step={this.step}
										value={value}
										onInput={this.onInputHandler}
										onBlur={this.onBlurHandler}
										onFocus={this.onFocusHandler}
										class={{
											'invalid': this.state === EValidationState.Invalid,
											'has-icon': !isEmpty(this.icon),
											'slotted': this.hasRightSlot,
											'forced-focus': this.focused
										}}
										readonly={this.uneditable}
									/>
									{this.icon && (
										<kv-icon
											name={this.icon}
											exportparts="icon"
											class={{
												invalid: this.state === EValidationState.Invalid,
												disabled: this.disabled,
												focus: this.focused
											}}
										/>
									)}
									{this.hasRightSlot && (
										<div class={{ 'right-slot-container': true, 'focus': this.focused }}>
											<slot name="right-slot"></slot>
										</div>
									)}
								</Fragment>
							)}
							{this.loading && <div class="input-container-loading"></div>}
						</div>
						<kv-form-help-text helpText={this._helpTexts} state={this.state}></kv-form-help-text>
					</div>
					{!isEmpty(this.examples) ? (
						<datalist id={`examples_${id}`}>
							{this.examples.map(example => (
								<option key={example} value={example}></option>
							))}
						</datalist>
					) : null}
				</kv-tooltip>
			</Host>
		);
	}
}
