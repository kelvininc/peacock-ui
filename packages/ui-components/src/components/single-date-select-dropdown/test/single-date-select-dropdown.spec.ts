import { SpecPage } from '@stencil/core/internal';
import { newSpecPage } from '@stencil/core/testing';
import { formatDatetime } from '../../../utils/date.helper';
import { KvTextField } from '../../text-field/text-field';
import { KvSingleDateSelectDropdown } from '../single-date-select-dropdown';

describe('Single Date Input (unit tests)', () => {
	let page: SpecPage;
	let component: KvSingleDateSelectDropdown;

	describe('when uses default props', () => {
		beforeEach(async () => {
			page = await newSpecPage({
				components: [KvSingleDateSelectDropdown, KvTextField],
				html: '<kv-single-date-select-dropdown></kv-single-date-select-dropdown>'
			});
			component = page.rootInstance;
		});

		it('should match the snapshot', () => {
			expect(page.root).toMatchSnapshot();
		});
	});

	describe('when `getFormattedSelectedDate` is called', () => {
		describe('and there is no selected date', () => {
			let result: string | undefined;

			beforeEach(async () => {
				page = await newSpecPage({
					components: [KvSingleDateSelectDropdown, KvTextField],
					html: '<kv-single-date-select-dropdown></kv-single-date-select-dropdown>'
				});
				component = page.rootInstance;
				result = component.getFormattedSelectedDate();
			});

			it('should return undefined', () => {
				expect(result).toBeUndefined();
			});
		});

		describe('and there is a selected date', () => {
			describe('and a custom date mask is defined', () => {
				let result: string | undefined;

				beforeEach(async () => {
					page = await newSpecPage({
						components: [KvSingleDateSelectDropdown, KvTextField],
						html: '<kv-single-date-select-dropdown selected-date="2022-04-01" date-mask="Do MMM, YYYY"></kv-single-date-select-dropdown>'
					});
					component = page.rootInstance;
					result = component.getFormattedSelectedDate();
				});

				it('should return an array with the selected date', () => {
					expect(result).toEqual('1st Apr, 2022');
				});
			});

			describe('and a custom date mask is not defined', () => {
				let result: string | undefined;

				beforeEach(async () => {
					page = await newSpecPage({
						components: [KvSingleDateSelectDropdown, KvTextField],
						html: '<kv-single-date-select-dropdown selected-date="2022-04-01"></kv-single-date-select-dropdown>'
					});
					component = page.rootInstance;
					result = component.getFormattedSelectedDate();
				});

				it('should return an array with the selected date', () => {
					expect(result).toEqual('2022-04-01 00:00:00');
				});
			});
		});
	});
});
