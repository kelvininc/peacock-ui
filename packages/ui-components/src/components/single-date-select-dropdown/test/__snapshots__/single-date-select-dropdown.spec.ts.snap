// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single Date Input (unit tests) when uses default props should match the snapshot 1`] = `
<kv-single-date-select-dropdown>
  <mock:shadow-root>
    <div class="single-date-select-dropdown">
      <kv-dropdown icon="kv-calendar" placeholder="Select a date" state="none" type="text">
        <div class="calendar-container">
          <kv-calendar-single-date-selector></kv-calendar-single-date-selector>
        </div>
      </kv-dropdown>
    </div>
  </mock:shadow-root>
</kv-single-date-select-dropdown>
`;
